waku/waku_mix/protocol.nim

```diff
@@ -0,0 +1,160 @@
+{.push raises: [].}
+
+import chronicles, std/[options, sequtils], chronos, results, metrics
+
+import
+  libp2p/crypto/curve25519,
+  mix/mix_protocol,
+  mix/mix_node,
+  mix/mix_metrics,
+  mix/tag_manager,
+  libp2p/[multiaddress, multicodec, peerid]
+
+import
+  ../node/peer_manager,
+  ../waku_core,
+  ../waku_enr/mix,
+  ../waku_enr,
+  ../node/peer_manager/waku_peer_store,
+  ../common/nimchronos
+
+logScope:
+  topics = "waku mix"
+
+type
+  WakuMix* = ref object of MixProtocol
+    peerManager*: PeerManager
+    clusterId: uint16
+
+  WakuMixResult*[T] = Result[T, string]
+
+proc mixPoolFilter*(cluster: Option[uint16], peer: RemotePeerInfo): bool =
+  # Note that origin based(discv5) filtering is not done intentionally
+  # so that more mix nodes can be discovered.
+  if peer.enr.isNone():
+    trace "peer has no ENR", peer = $peer
+    return false
+
+  if cluster.isSome() and peer.enr.get().isClusterMismatched(cluster.get()):
+    debug "peer has mismatching cluster", peer = $peer
+    return false
+
+  # Filter if mix is enabled
+  if not peer.enr.get().supportsCapability(Capabilities.Mix):
+    debug "peer doesn't support mix", peer = $peer
+    return false
+
+  return true
+
+proc appendPeerIdToMultiaddr*(multiaddr: MultiAddress, peerId: PeerId): MultiAddress =
+  if multiaddr.contains(multiCodec("p2p")).get():
+    return multiaddr
+
+  var maddrStr = multiaddr.toString().valueOr:
+    error "Failed to convert multiaddress to string.", err = error
+    return multiaddr
+  maddrStr.add("/p2p/" & $peerId)
+  var cleanAddr = MultiAddress.init(maddrStr).valueOr:
+    error "Failed to convert string to multiaddress.", err = error
+    return multiaddr
+  return cleanAddr
+
+proc populateMixNodePool*(mix: WakuMix) {.async.} =
+  # populate only peers that i) are reachable ii) share cluster iii) support mix
+  let remotePeers = mix.peerManager.switch.peerStore.getReachablePeers().filterIt(
+      mixPoolFilter(some(mix.clusterId), it)
+    )
+  var mixNodes = initTable[PeerId, MixPubInfo]()
+
+  for i in 0 ..< min(remotePeers.len, 100):
+    let remotePeerENR = remotePeers[i].enr.get()
```

Comment:
Mmm. I understand why we need it for now, but it would indeed be great to in future not have to know about `ENR` within the Waku Mix protocol